lightning/src/ln/package.rs

```diff
@@ -1,8 +1,719 @@
+		CounterpartyOfferedHTLCOutput {
+			per_commitment_point,
+			counterparty_delayed_payment_base_key,
+			counterparty_htlc_base_key,
+			preimage,
+			htlc
+		}
+	}
+}
+
+impl_writeable!(CounterpartyOfferedHTLCOutput, 33*3 + 32 + HTLC_OUTPUT_IN_COMMITMENT_SIZE, {
+	per_commitment_point,
+	counterparty_delayed_payment_base_key,
+	counterparty_htlc_base_key,
+	preimage,
+	htlc
+});
+
+/// A struct to describe a HTLC output on a counterparty commitment transaction.
+///
+/// HTLCOutputInCommitment (hash, timelock, directon) and pubkeys are used to generate a suitable
+/// witnessScript.
+#[derive(Clone, PartialEq)]
+pub(crate) struct CounterpartyReceivedHTLCOutput {
+	per_commitment_point: PublicKey,
+	counterparty_delayed_payment_base_key: PublicKey,
+	counterparty_htlc_base_key: PublicKey,
+	htlc: HTLCOutputInCommitment
+}
+
+impl CounterpartyReceivedHTLCOutput {
+	pub(crate) fn build(per_commitment_point: PublicKey, counterparty_delayed_payment_base_key: PublicKey, counterparty_htlc_base_key: PublicKey, htlc: HTLCOutputInCommitment) -> Self {
+		CounterpartyReceivedHTLCOutput {
+			per_commitment_point,
+			counterparty_delayed_payment_base_key,
+			counterparty_htlc_base_key,
+			htlc
+		}
+	}
+}
+
+impl_writeable!(CounterpartyReceivedHTLCOutput, 33*3 + HTLC_OUTPUT_IN_COMMITMENT_SIZE, {
+	per_commitment_point,
+	counterparty_delayed_payment_base_key,
+	counterparty_htlc_base_key,
+	htlc
+});
+
+/// A struct to describe a HTLC output on holder commitment transaction.
+///
+/// Either offered or received, the amount is always used as part of the bip143 sighash.
+/// Preimage is only included as part of the witness in former case.
+#[derive(Clone, PartialEq)]
+pub(crate) struct HolderHTLCOutput {
+	preimage: Option<PaymentPreimage>,
+	amount: u64,
+}
+
+impl HolderHTLCOutput {
+	pub(crate) fn build(preimage: Option<PaymentPreimage>, amount: u64) -> Self {
+		HolderHTLCOutput {
+			preimage,
+			amount
+		}
+	}
+}
+
+impl_writeable_len_match!(HolderHTLCOutput, {
+		{ HolderHTLCOutput { preimage: None, .. }, 1 + 8 },
+		{ _, 1 + 32 + 8 }
+	}, {
+	preimage,
+	amount
+});
+
+/// A struct to describe the channel output on the funding transaction.
+///
+/// witnessScript is used as part of the witness redeeming the funding utxo.
+#[derive(Clone, PartialEq)]
+pub(crate) struct HolderFundingOutput {
+	funding_redeemscript: Script,
+}
+
+impl HolderFundingOutput {
+	pub(crate) fn build(funding_redeemscript: Script) -> Self {
+		HolderFundingOutput {
+			funding_redeemscript,
+		}
+	}
+}
+
+impl_writeable_len_match!(HolderFundingOutput, {
+	{ HolderFundingOutput { ref funding_redeemscript }, 2 + funding_redeemscript.len() }
+	}, {
+	funding_redeemscript
+});
+
+/// A enum marking output category
+#[derive(PartialEq)]
+enum OutputCategory {
+	Revoked,
+	CounterpartyOfferedHTLC,
+	CounterpartyReceivedHTLC,
+	HolderHTLC,
+	HolderFunding,
+}
```

Comment:
Looking at how this is used, it doesn't really seem necessary. You can instead have a method on `PackageSolvingData` that takes another `PackageSolvingData` and returns whether they are compatible. No need to introduce the term "category" which isn't defined. This would also eliminate the redundancy between `OutputCategory` and `PackageSolvingData`.

Reply:
BTW, such comparison can be accomplished for complex enum types like `PackageSolvingData` using `std::mem::discriminant`.

https://doc.rust-lang.org/std/mem/fn.discriminant.html

Reply:
Okay finally dropped OutputCategory/OutputType! And leverage usage of `std::mem::discriminant`. There is still a bunch of special logic in `output_type_eq` to handle the forced-equality for revoked outputs.